[macOS] Stop copying image URL when using "Copy Image" context menu item

nt1m opened the pull request:
#### 8cbf542936d552ee9e32929e03410de79a255496
<pre>
[macOS] Stop copying image URL when using &quot;Copy Image&quot; context menu item
<a href="https://bugs.webkit.org/show_bug.cgi?id=301889">https://bugs.webkit.org/show_bug.cgi?id=301889</a>
<a href="https://rdar.apple.com/76598990">rdar://76598990</a>

Reviewed by Darin Adler.

Safari currently copies the image URL when using the &quot;Copy Image&quot; context menu item.
That results in extra items in the clipboard DataTransfer object. This is different from Chrome and Firefox which only copy the image.

This caused issues on Twitter, Facebook Messenger, Instagram and other sites, where pasting an image after copying it
from Safari would paste the image URL as text instead of uploading the image.

To fix this, we stop including the image URL in the DataTransfer object when using the &quot;Copy Image&quot; context menu item.

Test: http/tests/pasteboard/copy-image-from-context-menu.html

* LayoutTests/http/tests/pasteboard/copy-image-from-context-menu-expected.txt: Added.
* LayoutTests/http/tests/pasteboard/copy-image-from-context-menu.html: Added.
* LayoutTests/http/tests/pasteboard/resources/green-100.png: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/platform/mac/PasteboardMac.mm:
(WebCore::writableTypesForImage):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::copyImageAtLocation):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: <a href="https://commits.webkit.org/304166@main">https://commits.webkit.org/304166@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dfdd3356fadae697533228a14a95160b017d6899

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134759 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7203 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46007 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142279 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86680 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/9a814ee3-5675-45f9-a2e6-d40783eca09e) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136629 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7813 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7055 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/102989 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/70271 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/78bd8804-2fc4-4e21-af2d-41dc30e22d21) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137706 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5502 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120786 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83805 "Passed tests") | | ⏳ 🛠 vision-apple 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-26-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/2955 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2868 "Built successfully") | | 
| | [⏳ 🧪 api-ios](https://ews-build.webkit.org/#/builders/API-Tests-iOS-Simulator-EWS "Waiting to run tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/38933 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/144974 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6879 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39513 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/111379 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/6949 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5769 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111689 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28334 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5181 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117065 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60799 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/6927 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35248 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6726 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70508 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/6962 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6835 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

darinadler commented on LayoutTests/platform/ios/TestExpectations:
```diff
@@ -1022,6 +1022,9 @@ svg/text/selection-doubleclick.svg [ Skip ]
 svg/text/selection-tripleclick.svg [ Skip ]
 imported/w3c/web-platform-tests/css/css-shadow-parts/invalidation-part-pseudo.html [ Skip ]
 
+# iOS has a different copy image behavior.
```
I know you gave a reason for doing this only on Mac but I do not agree with this rationale. The Copy command also is almost certainly expected by almost all users to copy an image only, not a URL, on iOS, and I think we need to fix this there too. I believe I am the one who made copy commands silently copy the URL too because "why not?" and I do not think that decision of mine was a good decision. The unproven idea that some people depend on this on iOS seems extremely unlikely to me.

nt1m replied:
I'll write another PR for iOS shortly!

darinadler commented on Source/WebCore/testing/Internals.cpp:
```diff
@@ -8165,4 +8165,26 @@ bool Internals::isModelElementIntersectingViewport(HTMLModelElement& element)
 #endif
 
+#if !PLATFORM(IOS_FAMILY)
+ExceptionOr<void> Internals::copyImageAtLocation(int x, int y)
```
How will this link successfully on iOS? It is unconditional in the IDL and the class definition but then omitted here.

darinadler replied:
I see that iOS builds are failing, and I think the above is the reason why.

darinadler commented on Source/WebCore/platform/mac/PasteboardMac.mm:
```diff
@@ -78,7 +78,6 @@
 {
     Vector<String> types;
     types.append(String(legacyTIFFPasteboardTypeSingleton()));
-    types.appendVector(writableTypesForURL());
```
We should also remove the call to `writeURLForTypes`, which is no longer needed, instead of leaving in all that dead code.

nt1m replied:
I undid this, because this affects `event.clipboardData.types` in a way that makes it stop matching Chrome

darinadler reviewed (approved):
I'm glad we are tackling this but I don't think this is exactly right.

nt1m committed:
[macOS] Stop copying image URL when using "Copy Image" context menu item
https://bugs.webkit.org/show_bug.cgi?id=301889
rdar://76598990

Reviewed by Darin Adler.

Safari currently copies the image URL when using the "Copy Image" context menu item.
That results in extra items in the clipboard DataTransfer object. This is different from Chrome and Firefox which only copy the image.

This caused issues on Twitter, Facebook Messenger, Instagram and other sites, where pasting an image after copying it
from Safari would paste the image URL as text instead of uploading the image.

To fix this, we stop including the image URL in the DataTransfer object when using the "Copy Image" context menu item.

Test: http/tests/pasteboard/copy-image-from-context-menu.html

* LayoutTests/http/tests/pasteboard/copy-image-from-context-menu-expected.txt: Added.
* LayoutTests/http/tests/pasteboard/copy-image-from-context-menu.html: Added.
* LayoutTests/http/tests/pasteboard/resources/green-100.png: Added.
* LayoutTests/platform/ios/TestExpectations:
* LayoutTests/platform/mac-wk1/TestExpectations:
* Source/WebCore/platform/mac/PasteboardMac.mm:
(WebCore::writableTypesForImage):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::copyImageAtLocation):
* Source/WebCore/testing/Internals.h:
* Source/WebCore/testing/Internals.idl:

Canonical link: https://commits.webkit.org/304166@main
